### PR TITLE
codeintel: Reduce gitserver errors for missing revisions

### PR DESCRIPTION
**Pull request by efritz**
If we hit a revision not found error in the codeintel client, we do not want to treat these as _alertable_ errors as they are generally expected due to force-pushing over commits which codeintel holds references to.

We should move the handling of this class of errors up the stack. We still return the original error, we just don't emit a metric for it.

https://github.com/sourcegraph/sourcegraph/pull/16844 handles this error on the commit graph updater side. A future effort will need to do the same thing for the index schedulers.

### CODE REVIEW

File: enterprise/internal/codeintel/gitserver/client.go [hunk 1]
```diff
@@ -326,3 +326,16 @@ func (c *Client) repositoryIDToRepo(ctx context.Context, repositoryID int) (api.
 
 	return api.RepoName(repoName), nil
 }
+
+// with calls With on the given operation, but masks the error pointer in the case
+// that the error is due to an unknown revision. This function should be used over
+// calling With directly in this package whenever we use a revision that was stored
+// in a code intelligence-owned data store. These commits can fall out of sync with
+// gitserver and upstream callers are expected to handle these cases appropriately.
+func with(ctx context.Context, op *observation.Operation, err *error, args observation.Args) (context.Context, observation.FinishFunc) {
```
Comment (scalabilitysolved) [inline enterprise/internal/codeintel/gitserver/client.go:335]: Likely similar to my comment in the other PR, two concepts of `with/With` probably allow for future misunderstanding in usage.  Maybe call explicitly that it is `traceIgnoreRevisionErr` or something similar.  I am likely falling foul of Go naming though :)

**Review by scalabilitysolved — APPROVED** [review]
One small naming question, otherwise :+1:

File: enterprise/internal/codeintel/gitserver/client.go (hunk 1 above)
Comment (efritz) [reply enterprise/internal/codeintel/gitserver/client.go:335]: Yeah I just realized this was also a broken implementation so I'm going try something else that wraps the operation. Will re-ping when I push again.

File: enterprise/internal/codeintel/gitserver/observability.go [hunk 2]
```diff
@@ -30,6 +31,7 @@ func makeOperations(observationContext *observation.Context) *operations {
 			Name:         fmt.Sprintf("codeintel.gitserver.%s", name),
 			MetricLabels: []string{name},
 			Metrics:      metrics,
+			ErrorFilter:  gitserver.IsRevisionNotFound,
```
Comment (scalabilitysolved) [inline enterprise/internal/codeintel/gitserver/observability.go:34]: Definitely cleaner :+1:

**Review by scalabilitysolved — APPROVED** [review]
Change looks good :shipit:

**Review by Strum355 — APPROVED** [review]

**Pull request merged**